stdlib/public/core/Array.swift

```diff
@@ -1913,183 +1913,110 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+      return elements.description
     }
-    if rhs.base.count == 0 {
-      return lhs
+    @inlinable
+    public subscript(_ index: Int) -> Element {
+      if index < elements.count {
+        return elements[index]
+      }
+      return .zero
     }
-    return Array.DifferentiableView(zip(lhs.base, rhs.base).map(-))
   }
 
+  @available(*, deprecated, renamed: "TangentVector")
+  public typealias DifferentiableView = TangentVector
+
   @inlinable
-  public subscript(_ index: Int) -> Element {
-    if index < base.count {
-      return base[index]
-    } else {
-      return Element.zero
+  public mutating func move(along direction: TangentVector) {
+    for i in indices {
+      self[i].move(along: direction.elements[i])
     }
   }
 }
 
-/// Makes `Array` differentiable as the product manifold of `Element`
-/// multiplied with itself `count` times.
-extension Array : Differentiable where Element : Differentiable {
-  // In an ideal world, `TangentVector` would be `[Element.TangentVector]`.
-  // Unfortunately, we cannot conform `Array` to `AdditiveArithmetic` for
-  // `TangentVector` because `Array` already has a static `+` method with
-  // different semantics from `AdditiveArithmetic.+`. So we use
-  // `Array.DifferentiableView` for all these associated types.
-  public typealias TangentVector =
-    Array<Element.TangentVector>.DifferentiableView
-
-  public mutating func move(along direction: TangentVector) {
-    var view = DifferentiableView(self)
-    view.move(along: direction)
-    self = view.base
+extension Array.TangentVector: Equatable
+  where Array.Element.TangentVector: Equatable {
+  @inlinable
+  public static func == (lhs: Self, rhs: Self) -> Bool {
+    return lhs.elements == rhs.elements
   }
 }
 
-extension Array where Element : Differentiable {
-  public func _vjpSubscript(index: Int) ->
-    (Element, (Element.TangentVector) -> TangentVector)
-  {
-    func pullback(_ gradientIn: Element.TangentVector) -> TangentVector {
-      var gradientOut = Array<Element.TangentVector>(
-        repeating: .zero,
-        count: count)
-      gradientOut[index] = gradientIn
-      return TangentVector(gradientOut)
+extension Array where Element: Differentiable {
+  @inlinable
+  func _vjpSubscript(
+    index: Int
+  ) -> (Element, (Element.TangentVector) -> TangentVector) {
+    func pullback(_ v: Element.TangentVector) -> TangentVector {
+      var result = [Element.TangentVector](repeating: .zero, count: count)
+      result[index] = v
+      return TangentVector(result)
     }
     return (self[index], pullback)
   }
 
-  public static func _vjpPlus(_ lhs: [Element], _ rhs: [Element]) ->
-    ([Element], (TangentVector) -> (TangentVector, TangentVector)) {
-      func pullback(_ gradientIn: TangentVector) ->
-        (TangentVector, TangentVector) {
-        precondition(
-          gradientIn.base.count == lhs.count + rhs.count,
-          "+ should receive gradient with count equal to sum of operand " +
```

Comment:
Why not keep this extra precondition description?

Reply:
It is kept on line 2014. The error message is not super user-facing IMO, so I removed it for now.